internal/compiler/typeloader.rs

```diff
@@ -259,34 +292,57 @@ impl TypeLoader {
             })
     }
 
-    async fn ensure_document_loaded<'b>(
-        &'b mut self,
+    async fn ensure_document_loaded<'a: 'b, 'b>(
+        state: &'a RefCell<BorrowedTypeLoader<'a>>,
         file_to_import: &'b str,
         import_token: Option<NodeOrToken>,
-        diagnostics: &'b mut BuildDiagnostics,
+        mut import_stack: HashSet<PathBuf>,
     ) -> Option<PathBuf> {
-        let (path, is_builtin) = self.resolve_import_path(import_token.as_ref(), file_to_import);
+        let (path, builtin) =
+            { state.borrow().tl.resolve_import_path(import_token.as_ref(), file_to_import) };
 
         let path_canon = dunce::canonicalize(&path).unwrap_or_else(|_| path.to_owned());
 
-        if self.all_documents.docs.get(path_canon.as_path()).is_some() {
-            return Some(path_canon);
-        }
-
-        // Drop &self lifetime attached to is_builtin, in order to mutable borrow self below
-        let builtin = is_builtin.map(|s| s.to_owned());
-        let is_builtin = builtin.is_some();
-
-        if !self.all_documents.currently_loading.insert(path_canon.clone()) {
-            diagnostics
+        if !import_stack.insert(path_canon.clone()) {
+            state
+                .borrow_mut()
+                .diag
                 .push_error(format!("Recursive import of \"{}\"", path.display()), &import_token);
             return None;
         }
 
+        let is_loaded = core::future::poll_fn(|cx| {
+            let mut state = state.borrow_mut();
+            let all_documents = &mut state.tl.all_documents;
+            match all_documents.currently_loading.entry(path_canon.clone()) {
+                std::collections::hash_map::Entry::Occupied(mut e) => {
+                    e.get_mut().push(cx.waker().clone());
```

Comment:
I'm now wondering if that is correct.  We might end up inserting the same waker many times into the vector.

Reply:
Right, that seems like it could happen.